src/System.Net.Http/src/System/Net/Http/Managed/AuthenticationHandler.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    internal sealed class AuthenticationHandler : HttpMessageHandler
+    {
+        private readonly HttpMessageHandler _innerHandler;
+        private readonly bool _preAuthenticate;
+        private readonly ICredentials _credentials;
+
+        public AuthenticationHandler(bool preAuthenticate, ICredentials credentials, HttpMessageHandler innerHandler)
+        {
+            if (innerHandler == null)
+            {
+                throw new ArgumentNullException(nameof(innerHandler));
+            }
+
+            if (credentials == null)
+            {
+                throw new ArgumentNullException(nameof(credentials));
+            }
+
+            _preAuthenticate = preAuthenticate;
+            _credentials = credentials;
+            _innerHandler = innerHandler;
+        }
+
+        private bool TrySetBasicAuthToken(HttpRequestMessage request)
+        {
+            NetworkCredential credential = _credentials.GetCredential(request.RequestUri, "Basic");
+            if (credential == null)
+            {
+                return false;
+            }
+
+            request.Headers.Authorization = new AuthenticationHeaderValue("Basic", BasicAuthenticationHelper.GetBasicTokenForCredential(credential));
+            return true;
+        }
+
+        protected internal override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if (_preAuthenticate)
+            {
+                TrySetBasicAuthToken(request);
+            }
+
+            HttpResponseMessage response = await _innerHandler.SendAsync(request, cancellationToken);
```

Comment:
I have an issue with the code pattern of littering the codebase with lots of ConfigureAwait calls. The problem is if you need to make sure an asynchronous continuation runs on the thread pool (e.g. a non default synch context is set which might cause a deadlock), then you need to call ConfigureAwait on every awaitable call and all it takes is to miss using it in one place which only occasionally goes async to cause unpredictable deadlocks for some users. For WCF, I took a different approach. At each public api async entry point, I call a helper method which is awaitable. I check if there is a non-default synchronization context set or the current TaskScheduler is non-default and if so, I return an awaitable which forces continuation to happen on the thread pool. If everything is default, the returned awaitable says it completed synchronously so the calling method just continues like normal. We can only do this because we know that all our calls will always eventually go async. With the cost of calling ConfigureAwait repeatedly (top level entry method isn't sufficient, you have to do it all the way down the call stack until you are certain it would have gone async at least once by then), and the risk of bugs if you miss somewhere, I decided to always go async to the thread pool at the start and not have to worry about it again. We can only do this as we know we will always go async before the method is completed. Basically, I dislike ConfigureAwait :smile: